### DISCUSSION TITLE
Schema Changes From Laravel?

### DISCUSSION THREAD

**Discussion by x7ryan** [planetscale/discussion#379] [Q&A] [answered]
So I have a Laravel app, and I am unsure how to handle schema changes. Laravel has its own built in scripted migration process. 

But Planetscale has this "feature" that prevents schema changes directly on a "production branch". While I am sure that's useful in some instances, it's not in this one.  Is there a way to disable this limitation on a database so my Laravel app's migration script can just apply the schema changes directly to it's production branch?

If not, the only other way I can see would be to Write my own script to use the pscale CLI to create a branch, adjust the Laravel app's environment variable to build the connection so it connects to the newly created development branch then run Laravel's migrations then use the pscale CLI again to create and approve a deploy request. That seems overly complicated for no real benefit IMO. Please tell me this isn't how I have to do this.

**Comment by orware** [comment, marked as answer]
@x7ryan,

The situation you've described is a common adjustment concern when getting started with most frameworks, including Laravel.

For example, our own team has put together something internally that probably operates similarly to what you've described for our own Rails applications that use PlanetScale, and encouraging the creation of something similar for the Laravel community is something that's on our radar, if the community doesn't build it first!

We do have a public API that is due to be released fairly soon that will also help facilitate this a bit since it would help to avoid needing to have our `pscale` CLI present to allow those types of operations.

For myself, I previously came from an environment where making schema changes directly was pretty commonplace, as it likely is for many of the potential users that are interested in using PlanetScale for their current/future projects, so I understand how our [branching workflow](https://planetscale.com/docs/concepts/planetscale-workflow) could feel like a hindrance initially. 

But our workflow has been built primarily to help avoid the issues that eventually result when your database reaches a certain size. For example, once your database is large enough, making schema changes is no longer almost instant and can instead take a considerable amount of time to complete, locking the affected table and affecting your application. Our branching workflow, online schema migrations created via our deploy request process all work together to help prevent those types of issues for large databases, keeping developer velocity high since making schema changes don't have to be pushed out to a database team and scheduled to run at special times.

In other words, it definitely serves a purpose, but I also understand if it may be something you aren't immediately ready for. If that's the case, you can choose to keep your PlanetScale database branch as a development branch, which isn't as highly available as our production branches, but you would be allowed to make schema changes directly against them.

Even better, recently we added the ability for users to demote production branches back into development ones, so you could choose to use that option via our UI, apply your Laravel migrations, and then turn the branch back into a production one afterward.

The option to demote back into a development branch should be shown next to the **Connect** button when directly on your production branch's **Overview** page:
![image](https://user-images.githubusercontent.com/931650/215216165-ee5a1b50-7c12-40da-8f92-09963e895da1.png)

**Comment by x7ryan** [comment]
Thanks for the response. I think a Public API would be great. My biggest frustration with that process is it's reliance on the CLI as it is a common scenario in the Laravel community to run migrations from the production server. 

I will definitely keep an eye out for that Public API and when it drops I could make a package for Laravel to use it.

It seems less than ideal to not get any of the benefits of a production branch just because of this. The temporary demotion seems like a good middle ground, but that's a manual step. I would love to see demotion & promotion endpoints in the Public API. Then my package could also give demotion as an alternative because I think that solves another issue I see with branching is that it can't do any migration that touch data since data is separate per branch.

**Reply by orware** [reply]
@x7ryan,

The first iteration of the Public API was officially released earlier today!

Read more in the [official announcement post](https://planetscale.com/blog/introducing-planetscale-api-and-oauth-applications) or in our [API documentation](https://api-docs.planetscale.com/reference/getting-started-with-planetscale-api).

**Comment by x7ryan** [comment]
Just leaving a note here, for anyone who is looking for an automated script to manage this, I just released a package: https://github.com/x7media/laravel-planetscale
